### PR TITLE
[SPARK-25018] [Infra] Use `Co-authored-by` and `Signed-off-by` git trailer in `merge_spark_pr.py`

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

In [Linux community](https://git.wiki.kernel.org/index.php/CommitMessageConventions), `Co-authored-by` and `Signed-off-by` git trailer have been used for awhile. 

Until recently, Github adopted `Co-authored-by` to include the work of co-authors in the profile contributions graph and the repository's statistics. It's a convention for recognizing multiple authors, and can encourage people to collaborate in OSS communities.

Git provides a command line tools to read the metadata to know who commits the code to upstream, but it's not as easy as having `Signed-off-by` as part of the message so developers can find who is the relevant committers who can help with certain part of the codebase easier.

For a single author PR, I purpose to use `Authored-by` and `Signed-off-by`, so the message will look like

```
Authored-by: Author's name <author@example.com>
Signed-off-by: Committer's name <committer@example.com>
```

For a multi-author PR, I purpose to use `Lead-authored-by:` and `Co-authored-by:` for the lead author and co-authors. The message will look like

```
Lead-authored-by: Lead Author's name <leadauthor@example.com>
Co-authored-by: CoAuthor's name <coauthor@example.com>
Signed-off-by: Committer's name <committer@example.com>
```

It's also useful to include `Reviewed-by:` to give credits to the people who participate on the code reviewing. We can add this in the next iteration.

### CODE REVIEW

**Comment by dbtsai** [comment]
There are couple other common used git trailers used in linux community, https://git.wiki.kernel.org/index.php/CommitMessageConventions

We may consider to add `Signed-off-by:` and `Reviewed-by:` which can be helpful to track the code development.

File: dev/merge_spark_pr.py [hunk 1]
```diff
@@ -154,20 +154,22 @@ def merge_pr(pr_num, target_ref, title, body, pr_repo_desc):
...
-        committer_name = run_cmd("git config --get user.name").strip()
-        committer_email = run_cmd("git config --get user.email").strip()
         message = "This patch had conflicts when merged, resolved by\nCommitter: %s <%s>" % (
             committer_name, committer_email)
         merge_message_flags += ["-m", message]
 
     # The string "Closes #%s" string is required for GitHub to correctly close the PR
     merge_message_flags += ["-m", "Closes #%s from %s." % (pr_num, pr_repo_desc)]
 
+    authors = "\n".join(["Co-authored-by: %s" % a for a in distinct_authors])
```
Comment (dbtsai) [inline dev/merge_spark_pr.py]: For single author commit, should we do `Authored-by` instead of `Co-authored-by`?

**Comment by HyukjinKwon** [comment]
+1

**Comment by felixcheung** [comment]
Jenkins, retest this please

**Review by felixcheung — APPROVED** [review]
SGTM

File: dev/merge_spark_pr.py (hunk 1 above)
Comment (felixcheung) [reply dev/merge_spark_pr.py]: or leave it as before Author: to be safe?

Comment (srowen) [reply dev/merge_spark_pr.py]: Do these particular keys have a meaning to github or other tools? if "Author:" is what something is looking for, we'd want to leave it, I presume. But adding more lines for co-authors seems fine. I would have just listed multiple "Author:" lines, but sounds like there's another convention around "Co-authored-by:" that another tool looks for?

Comment (dbtsai) [reply dev/merge_spark_pr.py]: Seems that those keys are widely used in [ Linux community](https://git.wiki.kernel.org/index.php/CommitMessageConventions) for a long time, and Github adopted `Co-authored-by` to include the work of cu-authoers in the profile contributions graph and the repository's statistics.

This has to be in the end of the message, and Github will just ignore those keys that are not recognized. 

If we want to keep both multiple lines of `Author:` and `Co-authored-by:`,  the information will be duplicated twice. I would like to purpose to just keep `Co-authored-by:`. For single developer commit, we can just use `Authored-by:` for consistency.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@dbtsai, I sent a PR to your branch https://github.com/dbtsai/spark/pull/3 in case you want to test this and if I understood correctly.

**Comment by dbtsai** [comment]
@HyukjinKwon Thanks for pushing a test. I used this script to merge this PR into my own repo. Here is the result. https://github.com/dbtsai/spark/commits/PR_TOOL_MERGE_PR_21991_MASTER As you can see, three authors are properly shown in git status.

File: dev/merge_spark_pr.py [hunk 2]
```diff
@@ -142,6 +142,9 @@ def merge_pr(pr_num, target_ref, title, body, pr_repo_desc):
         distinct_authors[0])
     if primary_author == "":
         primary_author = distinct_authors[0]
+    else:
+        # When primary author is specified manually, put it at the head of author list.
+        distinct_authors.insert(0, primary_author)
```
Comment (dbtsai) [inline dev/merge_spark_pr.py]: @viirya good catch. When we put it at the head of the author list, do we need to dedupe it again?

Comment (viirya) [reply dev/merge_spark_pr.py]: I gave it a thought, but didn't do it. If you think it is necessary, I will push another commit for it.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
I am merging this since this is not actually tested and only thing is Python linter which is already passed.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dbtsai** [comment]
@HyukjinKwon thanks! Is it possible to use this script to merge this PR which has many people involve? A good demonstration of collaboration in the community.

**Comment by HyukjinKwon** [comment]
Yes, I think so. I was about to merge this in that way :-). Seems to me we are good to merge now since the current change is only checked by Python lint which is already passed.

**Comment by HyukjinKwon** [comment]
Let me leave it to you @dbtsai. I thought you live in a timezone completely different with me.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
https://github.com/apache/spark/commit/51bee7aca13451167fa3e701fcd60f023eae5e61 looks good :-)

**Comment by viirya** [comment]
The failed test is `FlatMapGroupsWithStateSuite.flatMapGroupsWithState`. I saw it fails some times occasionally. I think it should not be related to this change. @HyukjinKwon @dbtsai

**Comment by dbtsai** [comment]
Thanks all!

**Pull request closed without merging**